### PR TITLE
[SYCL][PI][CUDA] Implement profiling info queries

### PR DESCRIPTION
**Pull request by steffenlarsen**
Implements profiling info queries for PI_PROFILING_COMMAND_QUEUED,
PI_PROFILING_COMMAND_SUBMIT, and PI_PROFILING_COMMAND_START in
piEventGetProfilingInfo for the CUDA backend.

Signed-off-by: Steffen Larsen <steffen.larsen@codeplay.com>

### CODE REVIEW

File: sycl/include/CL/sycl/detail/pi.h [hunk 1]
```diff
@@ -365,6 +365,13 @@ typedef enum {
   PI_EVENT_INFO_CONTEXT = CL_EVENT_CONTEXT
 } _pi_event_info;
 
+typedef enum {
+  PI_PROFILING_COMMAND_QUEUED = CL_PROFILING_COMMAND_QUEUED,
```
Comment (smaslov-intel) [inline sycl/include/CL/sycl/detail/pi.h]: Please repeat enum type name as a prefix of enum values: PI_PROFILING_INFO_COMMAND_QUEUED

File: sycl/plugins/cuda/pi_cuda.cpp [hunk 2]
```diff
@@ -100,12 +100,13 @@ pi_result cuda_piEventRetain(pi_event event);
...
-      isRecorded_{false},
-      isStarted_{false}, event_{nullptr}, queue_{queue}, context_{context} {
+    : commandType_{type}, refCount_{1}, isCompleted_{false}, isRecorded_{false},
+      isStarted_{false}, evEnd_{nullptr}, evStart_{nullptr}, evQueued_{nullptr},
+      queue_{queue}, context_{context} {
 
   if (is_native_event()) {
-    PI_CHECK_ERROR(cuEventCreate(&event_, 0));
+    PI_CHECK_ERROR(cuEventCreate(&evQueued_, 0));
+    PI_CHECK_ERROR(cuEventCreate(&evEnd_, 0));
```
Comment (romanovvlad) [inline sycl/plugins/cuda/pi_cuda.cpp]: Shouldn't all this stuff be done only if queue has "enable_profiling" property?

File: sycl/plugins/cuda/pi_cuda.cpp [hunk 3]
```diff
@@ -131,6 +132,8 @@ pi_result _pi_event::start() {
 
   try {
     if (is_native_event()) {
+      // NOTE: This relies on the default queue to be unused.
+      result = PI_CHECK_ERROR(cuEventRecord(evQueued_, 0));
```
[the remark this answers is not in the document]
Comment (steffenlarsen) [reply sycl/plugins/cuda/pi_cuda.cpp]: The 0 specifies that we want to use the default stream in the active context. This is a very common behavior for CUDA operations that if they take a CUDA stream a 0 denotes the use of default stream.

I see that the comment is badly worded, so I'll adjust it slightly.

[the remark this answers is not in the document]
Comment (steffenlarsen) [reply sycl/plugins/cuda/pi_cuda.cpp]: However, your question still stands w.r.t. the `cuEventCreate` calls. I've changed them to use `CU_EVENT_DEFAULT` to make the intentions more clear.

File: sycl/plugins/cuda/pi_cuda.cpp (hunk 2 above)
Comment (steffenlarsen) [reply sycl/plugins/cuda/pi_cuda.cpp]: Good point! However, we still need to create and record `evEnd_` as it is also used for synchronization.

File: sycl/include/CL/sycl/detail/pi.h (hunk 1 above)
Comment (steffenlarsen) [reply sycl/include/CL/sycl/detail/pi.h]: Makes sense. It has been changed.

**Comment by steffenlarsen** [comment]
Resolved merge conflicts.

**Pull request merged**